sdk/storage/AzureStorageBlob/Source/DataStructures/StorageStreamUploader.swift

```diff
@@ -0,0 +1,540 @@
+// swiftlint:disable cyclomatic_complexity
+
+import AzureCore
+import Foundation
+
+/// Class used to upload an individual data chunk.
+internal class ChunkUploader {
+    // MARK: Properties
+
+    internal let blockId: String
+
+    internal let blobName: String
+
+    internal let containerName: String
+
+    internal let client: StorageBlobClient
+
+    internal let options: UploadBlobOptions
+
+    internal let uploadSource: URL
+
+    internal var streamStart: UInt64?
+
+    internal var startRange: Int
+
+    internal var endRange: Int
+
+    internal var isEncrypted: Bool {
+        return options.encryptionOptions?.key != nil || options.encryptionOptions?.keyResolver != nil
+    }
+
+    // MARK: Initializers
+
+    /// Creates a `ChunkUploader` object.
+    /// - Parameters:
+    ///   - blob: The name of the blob.
+    ///   - container: The name of the stoarge container in which to upload the blob.
+    ///   - client: The `StorageBlobClient` that initiated the request.
+    ///   - url: The source URL of the blob object.
+    ///   - blockId: A unique block identifer.
+    ///   - startRange: The start point, in bytes, of the upload request.
+    ///   - endRange: The end point, in bytes, of the upload request.
+    ///   - options: An `UploadBlobOptions` object with which to control the upload.
+    public init(
+        blob: String,
+        container: String,
+        blockId: String,
+        client: StorageBlobClient,
+        url: URL,
+        startRange: Int,
+        endRange: Int,
+        options: UploadBlobOptions
+    ) {
+        self.blockId = blockId
+        self.blobName = blob
+        self.containerName = container
+        self.client = client
+        self.options = options
+        self.uploadSource = url
+        self.startRange = startRange
+        self.endRange = endRange
+    }
+
+    // MARK: Public Methods
+
+    /// Begin the upload process.
+    /// - Parameters:
+    ///   - requestId: Unique request ID (GUID) for the operation.
+    ///   - completion: A completion handler that forwards the downloaded data.
+    public func upload(
+        requestId: String? = nil,
+        transactionalContentMd5: Data? = nil,
+        transactionalContentCrc64: Data? = nil,
+        then completion: @escaping HTTPResultHandler<Data>
+    ) {
+        // Construct URL
+        let urlTemplate = "/{container}/{blob}"
+        let pathParams = [
+            "container": containerName,
+            "blob": blobName
+        ]
+        let url = client.url(forTemplate: urlTemplate, withKwargs: pathParams)
+        let chunkSize = endRange - startRange
+        var buffer = Data(capacity: chunkSize)
+        do {
+            guard let fileHandle = try? FileHandle(forReadingFrom: uploadSource) else {
+                throw AzureError.general("Unable to open for reading: \(uploadSource)")
+            }
+            if #available(iOS 13.0, *) {
+                try fileHandle.seek(toOffset: UInt64(startRange))
```

Comment:
Is there not an equivalent of `try?` for methods that may throw but return void? Odd that we have to encase the thing in a `do... catch` block just because the method doesn't produce a return value.

Reply:
Yeah, let me take a look :)

Reply:
I took a look at this section and it's wrapped in a do/catch to catch various file-related errors that could occur. I updated the logic and removed the TODO. It now calls a `completion(.failure(error))` to ensure that any file errors are surfaced to the developer instead of being swallowed.